plugins/spectate.cpp

```diff
@@ -0,0 +1,157 @@
+
+DFhackCExport command_result plugin_init (color_ostream &out, std::vector <PluginCommand> &commands) {
+    commands.push_back(PluginCommand("spectate",
+                                     "Automated spectator mode.",
+                                     spectate,
+                                     false,
+                                     ""
+                                     " spectate\n"
+                                     "    toggles spectator mode\n"
+                                     "\n"));
+    return CR_OK;
+}
+
+DFhackCExport command_result plugin_shutdown (color_ostream &out) {
+    return CR_OK;
+}
+
+void onTick(color_ostream& out, void* tick);
+void onJobStart(color_ostream &out, void* job);
+void onJobCompletion(color_ostream &out, void* job);
+
+DFhackCExport command_result plugin_enable(color_ostream &out, bool enable) {
+    namespace EM = EventManager;
+    if (enable && !enabled) {
+        out.print("Spectate mode enabled!\n");
+        using namespace EM::EventType;
+        EM::EventHandler ticking(onTick, 15);
+        EM::EventHandler start(onJobStart, 0);
+        EM::EventHandler complete(onJobCompletion, 0);
+        EM::registerListener(EventType::TICK, ticking, plugin_self);
+        EM::registerListener(EventType::JOB_STARTED, start, plugin_self);
+        EM::registerListener(EventType::JOB_COMPLETED, complete, plugin_self);
+        //out.print("spectator mode enabled!\n");
+    } else if (!enable && enabled) {
+        out.print("Spectate mode disabled!\n");
+        EM::unregisterAll(plugin_self);
+        job_tracker.clear();
+        freq.clear();
+        //out.print("spectator mode disabled!\n");
+    } else {
+        return CR_FAILURE;
+    }
+    enabled = enable;
+    return CR_OK;
+}
+
+command_result spectate (color_ostream &out, std::vector <std::string> & parameters) {
+    return plugin_enable(out, !enabled);
+}
+
+void onTick(color_ostream& out, void* ptr){
+    int32_t tick = df::global::world->frame_counter;
+    if(!following_dwarf || (job_watched == nullptr && (tick - timestamp) > 50)) {
+        std::vector<df::unit*> dwarves;
+        for (size_t i = 0; i < df::global::world->units.active.size(); i++) {
+            df::unit* unit = world->units.active[i];
+            if (!Units::isCitizen(unit)) {
+                continue;
+            }
+            dwarves.push_back(unit);
+        }
+        std::uniform_int_distribution<> follow_any(0,dwarves.size()-1);
+        if (df::global::ui) {
+            df::unit* unit = dwarves[follow_any(RNG)];
+            df::global::ui->follow_unit = unit->id;
+            job_watched = unit->job.current_job;
+            following_dwarf = true;
+            if(!job_watched){
+                timestamp = tick;
+            }
+        }
+    }
+}
+
+void onJobStart(color_ostream& out, void* job_ptr) {
+    int32_t tick = df::global::world->frame_counter;
+    df::job* job = (df::job*) job_ptr;
+    int zcount = freq[job->pos.z]++;
```

Comment:
This is all random so this might not make a noticeable difference, but for an accurate current count I think this should use preincrement, not post.